src/main/java/redis/clients/jedis/MultiNodePipelineBase.java

```diff
@@ -58,14 +52,39 @@ public abstract class MultiNodePipelineBase implements PipelineCommands, Pipelin
   private final CommandObjects commandObjects;
   private GraphCommandObjects graphCommandObjects;
 
-  private final ExecutorService executorService = Executors.newFixedThreadPool(MULTI_NODE_PIPELINE_SYNC_WORKERS);
+  /**
+   * The following are the default parameters for the multi node pipeline executor
+   * Since Redis query is usually a slower IO operation (requires more threads),
+   * so we set DEFAULT_CORE_POOL_SIZE to be the same as the core
+   */
+  private static final long DEFAULT_KEEPALIVE_TIME_MS = 60000L;
+  private static final int DEFAULT_BLOCKING_QUEUE_SIZE = Protocol.CLUSTER_HASHSLOTS;
+  private static final int DEFAULT_CORE_POOL_SIZE = Runtime.getRuntime().availableProcessors();
+  private static final int DEFAULT_MAXIMUM_POOL_SIZE = Runtime.getRuntime().availableProcessors() * 2;
+  private static ExecutorService executorService = JedisThreadPoolBuilder.pool()
+      .setCoreSize(DEFAULT_CORE_POOL_SIZE)
+      .setMaxSize(DEFAULT_MAXIMUM_POOL_SIZE)
+      .setKeepAliveMillSecs(DEFAULT_KEEPALIVE_TIME_MS)
+      .setThreadNamePrefix("jedis-multi-node-pipeline")
+      .setWorkQueue(new ArrayBlockingQueue<>(DEFAULT_BLOCKING_QUEUE_SIZE)).build();
 
   public MultiNodePipelineBase(CommandObjects commandObjects) {
     pipelinedResponses = new LinkedHashMap<>();
     connections = new LinkedHashMap<>();
     this.commandObjects = commandObjects;
   }
 
+  /**
+   * Provide an interface for users to set executors themselves.
+   * @param executor the executor
+   */
+  public static void setExecutorService(ExecutorService executor) {
```

Comment:
IMO, I suggest that it is better to supply this parameter directly to the constructor method rather than the setter method.  : )

Reply:
@dengliming This is set once, works every time approach :)

Reply:
> IMO, I suggest that it is better to supply this parameter directly to the constructor method rather than the setter method. : )

We need to pass the executor when `new JedisCluster` or `new JedisSharding`, and need a clear name to avoid misunderstanding by users, and there are many modifications(JedisCluster->ClusterPipeline->MultiNodePipelineBase).

But the benefit is that we can create the executor lazily, if the user passes it, we use it; otherwise, we can create a static object.

But I personally prefer the current code, and it is one-time and clear to set directly through the `MultiNodePipelineBase#setExecutorService` interface (About cost: there is only a part of memory usage, if the user is not using it, no thread will be created).